clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -83,13 +87,71 @@ void BoundsAnalysis::InitInOutSets() {
   }
 }
 
+llvm::APSInt BoundsAnalysis::GetSwitchCaseVal(const Expr *CaseExpr) {
+  Expr *E = const_cast<Expr *>(CaseExpr);
+
+  while (auto *ICE = dyn_cast<ImplicitCastExpr>(E))
+    E = ICE->getSubExpr();
+
+  llvm::APSInt IntVal (Ctx.getTypeSize(Ctx.IntTy), 0);
+
+  // case '1': This is how it is represented in the CFG:
+  // ConstantExpr 'int'
+  //   -CharacterLiteral 'int' 49
+  // This is an IntegerConstantExpr. After the call to isIntegerConstantExpr,
+  // the variable IntVal would contain the value '1'.
+  if (E->isIntegerConstantExpr(IntVal, Ctx))
+    return IntVal;
+
+  // const int i = 1;
+  // case i: This is how it is represented in the CFG:
+  // ConstantExpr 'const int' lvalue 1
+  //   -DeclRefExpr 'const int' lvalue Var 'i' 'const int'
+  // This is not an IntegerConstantExpr. It is a ConstantExpr and we get its
+  // value by calling the getResultAsAPSInt() method.
+  if (const auto *CE = dyn_cast_or_null<ConstantExpr>(E))
+    return CE->getResultAsAPSInt();
```

Comment:
getResultAsAPSInt can only be called on a ConstantExpr. So I have to cast E to ConstantExpr here.